README
-*- markdown -*-
# Uridium

- www?

## DESCRIPTION:

FIX (describe your package)

## FEATURES

- Providers SDL Display, GDI and some drawing primitives to ruby, as well as 
  an eventloop with separate simulation and rendering objects and a mixer to 
  play sounds

## SYNOPSIS:

- Running asteroids.rb from examples/ will produce a classic 
  shoot-em-up game and highlight almost all features of this lib.

## REQUIREMENTS:

On linux you will need to install the following:
- libgl1-mesa-dev 
- libglu1-mesa-dev
- libsdl-dev
- libftgl-dev
- libsdl-mixer1.2-dev

- On debian/ubuntu it will be sufficient to do: 
   
  $ apt-get install libftgl-dev libsdl-mixer1.2-dev libsdl-dev</b>

- Will compile and run on Win32/MinGW, but no installation instructions are available at the moment.

- Darwin is not available at the moment.

## INSTALL:

   gem install

# LICENSE:

(The MIT License)

Copyright (c) 2009 FIX

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 0.2.0 / 2018-02-14

        * Migrate to SDL2
        * Ruby > 2.0
        * Support for Darwin

=== 0.1.64 / 2009-02-22

	* Using gemspec instead of hoe.


=== 0.1.63 / 2009-02-22

	* Corrected to compile with ruby 1.9.1


=== 0.1.62 / 2009-02-22

	* Added hoe generation files + Rakefile


=== 0.1.61 / 2008-07-24

	* Compiles on DARWIN/mac os x 10.5, but dumps core.


=== 0.1.60 / 2008-03-25

	* [svn] 	moar like 1.8.6


=== 0.1.59 / 2008-03-24

	* [svn] - Added starfield background.


=== 0.1.58 / 2008-03-22

	* [svn] - Asteroids playable.


=== 0.1.57 / 2008-03-18

	* [svn] 	Added destroy_sound.


=== 0.1.56 / 2008-03-17

	* [svn] - Cleaned up code.


=== 0.1.55 / 2008-03-17

	* [svn] 	HOT TIP:


=== 0.1.54 / 2008-03-17

	* [svn] 	Partial fixes ( none ) to broken eventloop on unix.


=== 0.1.53 / 2008-03-16

	* [svn] 	Added Sound as separate class from mixer.


=== 0.1.52 / 2008-03-16

	* [svn] - Rewrote most of Asteroids.


=== 0.1.51 / 2008-03-16

	* [svn] - Added Clock class for high resolution timing purposes.


=== 0.1.50 / 2008-03-10

	* [svn] 	Fire dem LAZERS! added play_sound to mixer. No unallocation or


=== 0.1.49 / 2008-03-07

	* [svn] 	Added draw_polygon_2d


=== 0.1.48 / 2008-03-07

	* [svn] - Added simple Asteroids example (wip).


=== 0.1.47 / 2008-02-26

	* [svn] - Optimized event loop polling out if event mask is not set.


=== 0.1.46 / 2008-02-26

	* [svn] - Added event dispatching for key and activation events.


=== 0.1.45 / 2008-02-24

	* [svn] - Fixed super class of Mixer.


=== 0.1.44 / 2008-02-24

	* [svn] - Fixed Mingw build with SDL_mixer.


=== 0.1.43 / 2008-02-24

	* [svn] 	Initial ( empty ) sound mixer.


=== 0.1.42 / 2008-02-23

	* [svn] - Added ActivityEvent.


=== 0.1.41 / 2008-02-23

	* [svn] - Removed the obsolete list of objects to compile.


=== 0.1.40 / 2008-02-23

	* [svn] - Cleaned up code.


=== 0.1.39 / 2008-02-23

	* [svn] - Cleaned up.


=== 0.1.38 / 2008-02-23

	* [svn] - Made extern "C" declarations more consistent.


=== 0.1.37 / 2008-02-22

	* [svn] - Added Event class.


=== 0.1.36 / 2008-02-20

	* [svn] - Added simple Event class for polling events.


=== 0.1.35 / 2008-02-20

	* [svn] - Cleaned up code.


=== 0.1.34 / 2008-02-20

	* [svn] - Moved cpp class identifiers to uridium.h.


=== 0.1.33 / 2008-02-20

	* [svn] - Tried to fix $LOAD_PATH in examples.


=== 0.1.32 / 2008-02-17

	* [svn] - Cleaned up code.


=== 0.1.31 / 2008-02-17

	* [svn] - Made the logo movement FPS independent.


=== 0.1.30 / 2008-02-17

	* [svn] - Added trans0, a method that resets the transform.


=== 0.1.29 / 2008-02-17

	* [svn] - Fixed gdi/rotate_z.


=== 0.1.28 / 2008-02-17

	* [svn] - Changed Ruby 1.9 only lambda to 1.8 style.


=== 0.1.27 / 2008-02-17

	* [svn] - Added idle_time.


=== 0.1.26 / 2008-02-16

	* [svn] - Fixed gdi/clear.


=== 0.1.25 / 2008-02-16

	* [svn] - Added ext, lib and examples.


=== 0.1.24 / 2008-02-12

	* [svn] 	- added gdi.draw_line.


=== 0.1.23 / 2008-02-12

	* [svn] 	Testing rotate_z


=== 0.1.22 / 2008-02-12

	* [svn] 	-added more fun to highcaliber.


=== 0.1.21 / 2008-02-12

	* [svn] 	Added defines for ruby < 1.8.6


=== 0.1.20 / 2008-02-12

	* [svn] - Removed unnecessary files (Makefile, nbproject, uridium.so).


=== 0.1.19 / 2008-02-12

	* [svn] - Added a precompiled Freetype dll for easier testing.


=== 0.1.18 / 2008-02-12

	* [svn] - Fixed the shared linking flags for Mingw.


=== 0.1.17 / 2008-02-12

	* [svn] - Fixed font rendering.


=== 0.1.16 / 2008-02-11

	* [svn] - Added a simple gdi demo.


=== 0.1.15 / 2008-02-11

	* [svn] - Nuked build.cmd, as it was only useful for MS compilers.


=== 0.1.14 / 2008-02-11

	* [svn] - Made Uridium Mingw32 compatible.


=== 0.1.13 / 2008-02-11

	* [svn]


=== 0.1.12 / 2008-02-11

	* [svn] uh-oh


=== 0.1.11 / 2008-02-11

	* [svn] Pitää poistaa display.rb että tätä voi käyttää.


=== 0.1.10 / 2008-02-04

	* [svn]


=== 0.1.9 / 2008-02-04

	* [svn] Wrapped font to a struct


=== 0.1.8 / 2008-01-23

	* [svn] cpp implementation of fonts. Dumps core though.


=== 0.1.7 / 2008-01-23

	* [svn]


=== 0.1.6 / 2008-01-23

	* [svn]


=== 0.1.5 / 2008-01-22

	* [svn] Added c++ implementation of gdi


=== 0.1.4 / 2008-01-22

	* [svn] stuff++


=== 0.1.3 / 2008-01-22

	* [svn] Add CC change for extension.


=== 0.1.2 / 2008-01-22

	* [svn] extconf for linus.


=== 0.1.1 / 2008-01-20

	* [svn] Almost compiles in unix, a linker error though.


=== 0.1.0 / 2008-01-19

	* [svn] Initial import


gdi.cpp
/* -*-c++-*-
 *
 * gdi.cpp
 *
 * Created on 22.1.2008, 22:07:18
 *
 */
#include "display.h"
#include "gdi.h"
#include <SDL/SDL.h>
#include <GL/gl.h>

extern "C" VALUE gdi_init_impl(VALUE self, VALUE display)
{
    rb_iv_set(self, "@display", display);
    return self;
}

/*
 *   call-seq: scale() => #
 *
 */
extern "C" VALUE gdi_scale_impl(VALUE self, VALUE val)
{
    double s = NUM2DBL(val);
    glScalef(s, s, s);
    return Qnil;
}


/*
 *   call-seq: clear() => #
 *
 */
extern "C" VALUE gdi_clear_impl(int argc, VALUE* argv, VALUE self)
{
     VALUE rb_color;
     unsigned int color = 0x00000000;
     rb_scan_args(argc, argv, "01", &rb_color);
     if (!NIL_P(color))
     {
       color = NUM2UINT(rb_color);
     }
     
     glClearColor((color & 0x00ff0000) >> 16, (color & 0x0000ff00) >> 8, (color & 0x000000ff), (color & 0xff000000) >> 24);
     glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
     return Qnil;
}


/*
 *   call-seq: flip() => #
 *
 */
extern "C" VALUE gdi_flip_impl(VALUE self)
{
    SDL_GL_SwapBuffers();
    return Qnil;
}

/*
 *   call-seq: rotate(int angle) => #
 *       Rotate object around by <b>angle</b>.
 *
 */
extern "C" VALUE gdi_rotate_z_impl(VALUE self, VALUE val)
{
    int a = NUM2INT(val);
    glRotatef(a, 0, 0, 1);
    return Qnil;
}


static VALUE rb_gdi;

void init_gdi()
{
  rb_gdi = rb_define_class("Gdi", rb_cObject);
  rb_define_method(rb_gdi, "initialize", (ruby_method*) &gdi_init_impl, 1);
  rb_define_method(rb_gdi,"scale",(ruby_method*) &gdi_scale_impl, 1);
  rb_define_method(rb_gdi,"clear",(ruby_method*) &gdi_clear_impl, -1);
  rb_define_method(rb_gdi,"rotate_z",(ruby_method*) &gdi_rotate_z_impl, 1);
  rb_define_method(rb_gdi,"flip",(ruby_method*) &gdi_flip_impl, 0);
}
